tree/dataframe/src/RNTupleDS.cxx

```diff
@@ -33,31 +33,83 @@
+
+/// An artifical field that transforms an RNTuple column that contains the offset of collections into
+/// collection sizes. It is used to provide the "number of" RDF columns for collections, e.g.
+/// `#jets` for a collection named `jets`.
+class RRDFCardinalityField : public ROOT::Experimental::Detail::RFieldBase {
+public:
+   ROOT::Experimental::RField<ClusterSize_t> fOffsetField;
+
+   static std::string TypeName() { return "ROOT::Experimental::ClusterSize_t::ValueType"; }
+   RRDFCardinalityField()
+     : Detail::RFieldBase("", TypeName(), ENTupleStructure::kLeaf, false /* isSimple */)
+     , fOffsetField("")
+   {
+   }
+   RRDFCardinalityField(RRDFCardinalityField&& other) = default;
+   RRDFCardinalityField& operator =(RRDFCardinalityField&& other) = default;
+   ~RRDFCardinalityField() = default;
+   std::unique_ptr<Detail::RFieldBase> Clone(std::string_view /* newName */) const final {
+      return std::make_unique<RRDFCardinalityField>();
+   }
+
+   /// Being virtual, the field doesn't have any columns of its own but it uses, indirectly, the columns
+   /// of fOffsetField when reading
+   void GenerateColumnsImpl() final { }
+
+   ROOT::Experimental::Detail::RFieldValue GenerateValue(void* where) final
+   {
+      return Detail::RFieldValue(this, static_cast<ClusterSize_t *>(where));
+   }
+   Detail::RFieldValue CaptureValue(void *where) final
+   {
+      return Detail::RFieldValue(true /* captureFlag */, this, where);
+   }
+   size_t GetValueSize() const final { return sizeof(ClusterSize_t); }
+
+   /// Get the number of elements of the collection identified by globalIndex
+   void ReadGlobalImpl(ROOT::Experimental::NTupleSize_t globalIndex,
+                       ROOT::Experimental::Detail::RFieldValue *value) final
+   {
+      RClusterIndex collectionStart;
+      fOffsetField.GetCollectionInfo(globalIndex, &collectionStart, value->Get<ClusterSize_t>());
+   }
+
+   /// Get the number of elements of the collection identified by clusterIndex
+   void ReadInClusterImpl(const ROOT::Experimental::RClusterIndex &clusterIndex,
+                          ROOT::Experimental::Detail::RFieldValue *value) final
+   {
+      RClusterIndex collectionStart;
+      fOffsetField.GetCollectionInfo(clusterIndex, &collectionStart, value->Get<ClusterSize_t>());
+   }
+};
+
+
+/// Every RDF column is represented by exactly one RNTuple field
+class RNTupleColumnReader : public ROOT::Detail::RDF::RColumnReaderBase {
+protected:
    using RFieldBase = ROOT::Experimental::Detail::RFieldBase;
    using RFieldValue = ROOT::Experimental::Detail::RFieldValue;
    using RPageSource = ROOT::Experimental::Detail::RPageSource;
 
-   std::unique_ptr<RFieldBase> fField;
-   RFieldValue fValue;
+   std::unique_ptr<RFieldBase> fField; ///< The field backing the RDF column
+   std::vector<DescriptorId_t> fSkeinIDs; ///< For inner collections, the field IDs of the outer collections
```

Comment:
```suggestion
   /// For inner collections, the field IDs of the outer collections, from outermost to innermost. Empty otherwise.
   std::vector<DescriptorId_t> fSkeinIDs;
```